### PR TITLE
Move the Route Registering to the Controllers

### PR DESCRIPTION
**Pull request by joehoyle**

### CODE REVIEW

**Comment by joehoyle** [comment]
@WP-API/amigos #reviewmerge

File: lib/endpoints/class-wp-json-posts-controller.php [hunk 1]
```diff
@@ -9,6 +9,76 @@ public function __construct( $post_type ) {
 	}
 
 	/**
+	 * Register the routes for the objects of the controller
+	 * 
+	 */
+	public function register_routes() {
+
+		parent::register_routes();
```
Comment (danielbachhuber) [inline lib/endpoints/class-wp-json-posts-controller.php]: If someone misses this call, the sky falls on their head?

File: lib/endpoints/class-wp-json-post-types-controller.php [hunk 2]
```diff
@@ -3,6 +3,33 @@
...
+
+		register_json_route( 'wp', '/types/schema', array(
+			'methods'         => WP_JSON_Server::READABLE,
+			'callback'        => array( $this, 'get_item_schema' ),
+		) );
+
+		register_json_route( 'wp', '/types/(?P<type>[\w-]+)', array(
+			'methods'         => WP_JSON_Server::READABLE,
+			'callback'        => array( $this, 'get_item' ),
+		) );
```
Comment (danielbachhuber) [inline lib/endpoints/class-wp-json-post-types-controller.php]: Should we make each of these register instances separate callbacks, so someone could subclass and override? How would someone go about disabling the registry of a route?

File: lib/endpoints/class-wp-json-posts-controller.php (hunk 1 above)
Comment (danielbachhuber) [inline lib/endpoints/class-wp-json-posts-controller.php]: Do we even want them to necessarily re-register the parent's routes?

Comment (joehoyle) [inline lib/endpoints/class-wp-json-posts-controller.php]: Perhaps not actually

File: lib/endpoints/class-wp-json-post-types-controller.php [hunk 3]
```diff
@@ -3,6 +3,33 @@
 class WP_JSON_Post_Types_Controller extends WP_JSON_Controller {
 
 	/**
+	 * Register the routes for the objects of the controller
+	 * 
+	 */
+	public function register_routes() {
+
+		parent::register_routes();
```
Comment (rmccue) [inline lib/endpoints/class-wp-json-post-types-controller.php]: Let's remove this unless it's needed. The base controller shouldn't need to register endpoints, right?

File: lib/endpoints/class-wp-json-posts-controller.php (hunk 1 above)
Comment (rmccue) [inline lib/endpoints/class-wp-json-posts-controller.php]: Let's remove this unless it's needed. The base controller shouldn't need to register endpoints, right?

Comment (joehoyle) [inline lib/endpoints/class-wp-json-posts-controller.php]: The idea was more to potentially have some DRY code in the base for things like schema endpoints, but after thinking on that I don't think we can do it automajically very well anyway

File: lib/endpoints/class-wp-json-post-types-controller.php (hunk 2 above)
Comment (rmccue) [inline lib/endpoints/class-wp-json-post-types-controller.php]: I think it's fine as-is here, but we probably need a `deregister_json_route` as well.

One thing I did like from v1 was that you took the `$routes` array and returned your modified version (filter-style), which meant you could `unset` anything you didn't want. Don't know how that would work with `register_json_route` though.

File: lib/endpoints/class-wp-json-comments-controller.php [hunk 4]
```diff
@@ -6,6 +6,123 @@
 class WP_JSON_Comments_Controller extends WP_JSON_Controller {
 
 	/**
+	 * Register the routes for the objects of the controller
+	 *
```
Comment (rmccue) [inline lib/endpoints/class-wp-json-comments-controller.php]: Extra line in the docblock should be trimmed here, and the description should have a trailing period.

File: lib/endpoints/class-wp-json-post-types-controller.php (hunk 2 above)
Comment (joehoyle) [inline lib/endpoints/class-wp-json-post-types-controller.php]: Created #972 for the deregistering

**Comment by joehoyle** [comment]
:clap:

**Pull request merged**